core/src/app.rs

```diff
@@ -211,16 +212,24 @@ impl App {
                 }
             }
 
-            let block_type_name = (block.block_type(), name.clone());
-            match block_type_names.contains(&block_type_name) {
-                true => Err(anyhow!(
-                    "Repeated block `{} {}`",
-                    block_type_name.0.to_string(),
-                    block_type_name.1
-                ))?,
-                false => {
-                    block_type_names.insert(block_type_name);
+            // check that blocks don't share the same name, except for `map/reduce` and `while/end`
+            if let Some(block_types) = block_types_by_name.get_mut(name) {
+                if block_types.len() > 1
+                    || !((block.block_type() == BlockType::End
+                        && block_types.contains(&BlockType::While))
+                        || (block.block_type() == BlockType::Reduce
+                            && block_types.contains(&BlockType::Map)))
+                {
+                    Err(anyhow!(
```

Comment:
Woow that's hard to parse.

```
if block_types.len() > 0 && 
  !(block.block_type() == BlockType::End && block_types.contains(&BlockType::While)) &&
  !(block.block_type() == BlockType::Reduce && block_types.contains(&BlockType::Map))
```
Seems easier to parse. The `> 1` is quite confusing because it makes it unclear what happens for the simple case of a block having the same name as a previous one?

Reply:
I still don't see in the current code path how == 0 is handled :grimacing:

Reply:
If it's 0, then `let Some(block_types) = block_types_by_name.get_mut(name)` cannot be true

Reply:
Added some comments let me know if that looks OK to you